File: java/src/com/helgeeichhorn/icatt/Lambert.java
package com.helgeeichhorn.icatt;

import org.apache.commons.math3.geometry.euclidean.threed.Vector3D;
import org.apache.commons.math3.special.Gamma;

public class Lambert {
    public static final class LambertResult {
        public Vector3D v0;
        public Vector3D v;

        LambertResult(Vector3D v0, Vector3D v) {
            this.v0 = v0;
            this.v = v;
        }
    }
    public static LambertResult solve(double k, Vector3D r0, Vector3D r, double tof) {
        boolean shortway = true;
        int numiter = 35;
        double rtol = 1e-8;
        return solve(k, r0, r, tof, shortway, numiter, rtol);
    }
    public static LambertResult solve(double k, Vector3D r0, Vector3D r, double tof,
                                 boolean shortway, int numiter, double rtol) {
        int tm;
        if (shortway) {
            tm = 1;
        } else {
            tm = -1;
        }
        double normR0 = r0.getNorm();
        double normR = r.getNorm();
        double cosDnu = r0.dotProduct(r) / (normR0*normR);
        double A = tm * Math.sqrt(normR * normR0 * (1 + cosDnu));

        if (A == 0.0) {
            throw new RuntimeException("Cannot compute orbit, phase angle is 180 degrees");
        }
        double psi = 0.0;
        double psiLow = -4*Math.PI;
        double psiUp = 4*Math.PI;

        double y = 0;
        int count = 0;
        while (count < numiter) {
            y = normR0 + normR + A * (psi * c3(psi) - 1) / Math.sqrt(c2(psi));
            if (A > 0 & y < 0) {
                while (y < 0) {
                    psiLow = psi;
                    psi = (0.8 * (1.0 / c3(psi)) *
                            (1.0 - (normR0 + normR) * Math.sqrt(c2(psi)) / A));
                    y = normR + normR0 + A * (psi * c3(psi) - 1) / Math.sqrt(c2(psi));
                }
            }
            double xi = Math.sqrt(y / c2(psi));
            double tofNew = (Math.pow(xi, 3) * c3(psi) + A * Math.sqrt(y)) / Math.sqrt(k);

            if (Math.abs((tofNew - tof) / tof) < rtol) {
                break;
            } else {
                count += 1;
                if (tofNew <= tof) {
                    psiLow = psi;
                } else {
                    psiUp = psi;
                }
                psi = (psiUp + psiLow) / 2;
            }
        }
        if (count >= numiter) {
            throw new RuntimeException("Maximum number of iterations reached.");
        }
        double f = 1 - y / normR0;
        double g = A * Math.sqrt(y / k);
        double gdot = 1 - y / normR;
        Vector3D v0 = r.subtract(r0.scalarMultiply(f)).scalarMultiply(1/g);
        Vector3D v = r.scalarMultiply(gdot).subtract(r0).scalarMultiply(1/g);
        return new LambertResult(v0, v);
    }

    public static double c2(double psi) {
        double res;
        double eps = 1.0;
        if (psi > eps) {
            res = (1 - Math.cos(Math.sqrt(psi))) / psi;
        } else if (psi < -eps) {
            res = (Math.cosh(Math.sqrt(-psi)) - 1) / -psi;
        } else {
            res = 1.0 / 2.0;
            double delta = (-psi) / Gamma.gamma(2+2+1);
            int k = 1;
            while (res + delta != res) {
                res += delta;
                k += 1;
                delta = Math.pow(-psi, k) / Gamma.gamma(2 * k + 2 + 1);
            }
        }
        return res;
    }

    public static double c3(double psi) {
        double res;
        double eps = 1.0;
        if (psi > eps) {
            res = (Math.sqrt(psi) - Math.sin(Math.sqrt(psi))) / (psi * Math.sqrt(psi));
        } else if (psi < -eps) {
            res = (Math.sinh(Math.sqrt(-psi)) - Math.sqrt(-psi)) / (-psi * Math.sqrt(-psi));
        } else {
            res = 1.0/6.0;
            double delta = (-psi) / Gamma.gamma(2 + 3 + 1);
            int k = 1;
            while (res + delta != res) {
                res += delta;
                k += 1;
                delta = Math.pow(-psi, k) / Gamma.gamma(2 * k + 3 + 1);
            }
        }
        return res;
    }
}

File: python/icatt/lambert.py
import time
import numpy as np
from math import gamma

from numba import jit, njit

def benchmark(times):
    r0 = np.array([5000.0, 10000.0, 2100.0])
    r = np.array([-14600.0, 2500.0, 7000.0])
    tof = 3600.0
    mu = 3.986004418e5
    lambert(mu, r0, r, tof)

    best = np.inf
    worst = -np.inf
    total = 0.0
    for _ in range(times):
        t0 = time.clock()
        lambert(mu, r0, r, tof)
        t1 = time.clock()
        current = t1 - t0
        if current < best:
            best = current
        if current > worst:
            worst = current
        total += current
    print("[",total/times,",",best,",",worst,"]")

@njit
def lambert(k, r0, r, tof, short=True, numiter=35, rtol=1e-8):
    if short:
        t_m = +1
    else:
        t_m = -1

    norm_r0 = np.dot(r0, r0)**.5
    norm_r = np.dot(r, r)**.5
    cos_dnu = np.dot(r0, r) / (norm_r0 * norm_r)

    A = t_m * (norm_r * norm_r0 * (1 + cos_dnu))**.5

    if A == 0.0:
        raise RuntimeError("Cannot compute orbit, phase angle is 180 degrees")

    psi = 0.0
    psi_low = -4 * np.pi
    psi_up = 4 * np.pi

    count = 0
    while count < numiter:
        y = norm_r0 + norm_r + A * (psi * c3(psi) - 1) / c2(psi)**.5
        if A > 0.0 and y < 0.0:
            # Readjust xi_low until y > 0.0
            # Translated directly from Vallado
            while y < 0.0:
                psi_low = psi
                psi = (0.8 * (1.0 / c3(psi)) *
                       (1.0 - (norm_r0 + norm_r) * np.sqrt(c2(psi)) / A))
                y = norm_r0 + norm_r + A * (psi * c3(psi) - 1) / c2(psi)**.5

        xi = np.sqrt(y / c2(psi))
        tof_new = (xi**3 * c3(psi) + A * np.sqrt(y)) / np.sqrt(k)

        # Convergence check
        if np.abs((tof_new - tof) / tof) < rtol:
            break
        else:
            count += 1
            # Bisection check
            if tof_new <= tof:
                psi_low = psi
            else:
                psi_up = psi
            psi = (psi_up + psi_low) / 2
    else:
        raise RuntimeError("Maximum number of iterations reached")

    f = 1 - y / norm_r0
    g = A * np.sqrt(y / k)

    gdot = 1 - y / norm_r

    v0 = (r - f * r0) / g
    v = (gdot * r - r0) / g

    return v0, v

@njit('f8(f8)')
def c2(psi):
    r"""Second Stumpff function.
    For positive arguments:
    .. math::
        c_2(\psi) = \frac{1 - \cos{\sqrt{\psi}}}{\psi}
    """
    eps = 1.0
    if psi > eps:
        res = (1 - np.cos(np.sqrt(psi))) / psi
    elif psi < -eps:
        res = (np.cosh(np.sqrt(-psi)) - 1) / (-psi)
    else:
        res = 1.0 / 2.0
        delta = (-psi) / gamma(2 + 2 + 1)
        k = 1
        while res + delta != res:
            res = res + delta
            k += 1
            delta = (-psi) ** k / gamma(2 * k + 2 + 1)

    return res


@njit('f8(f8)')
def c3(psi):
    r"""Third Stumpff function.
    For positive arguments:
    .. math::
        c_3(\psi) = \frac{\sqrt{\psi} - \sin{\sqrt{\psi}}}{\sqrt{\psi^3}}
    """
    eps = 1.0
    if psi > eps:
        res = (np.sqrt(psi) - np.sin(np.sqrt(psi))) / (psi * np.sqrt(psi))
    elif psi < -eps:
        res = (np.sinh(np.sqrt(-psi)) - np.sqrt(-psi)) / (-psi * np.sqrt(-psi))
    else:
        res = 1.0 / 6.0
        delta = (-psi) / gamma(2 + 3 + 1)
        k = 1
        while res + delta != res:
            res = res + delta
            k += 1
            delta = (-psi) ** k / gamma(2 * k + 3 + 1)

    return res
